File: everyshowing/src/main/java/com/wghcwc/everyshowing/activitystack/ActivityLifecycle.java
package com.wghcwc.everyshowing.activitystack;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.WeakHashMap;

/**
 * @author wghcwc
 * @date 19-12-5
 */
public class ActivityLifecycle implements Application.ActivityLifecycleCallbacks {
    public static ActivityLifecycle instance;

    private final WeakHashMap<Activity, LinkedList<ActivityChangeListener>> activityLinked;

    static {
        instance = new ActivityLifecycle();
    }

    private ActivityLifecycle() {
        activityLinked = new WeakHashMap<>();
    }

    public static ActivityLifecycle getInstance() {
        return instance;
    }

    public void add(Activity activity, ActivityChangeListener listener) {
        LinkedList<ActivityChangeListener> linkedList = activityLinked.get(activity);
        if (linkedList == null) {
            linkedList = new LinkedList<>();
        }
        linkedList.push(listener);
        activityLinked.put(activity, linkedList);

    }

    public void remove(Activity activity, ActivityChangeListener listener) {
        LinkedList<ActivityChangeListener> linkedList = activityLinked.get(activity);
        if (linkedList != null) {
            linkedList.remove(listener);
            if (linkedList.size() == 0) {
                activityLinked.remove(activity);
            }
        }
    }

    public void remove(Activity activity) {
        activityLinked.remove(activity);

    }

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        ActivityStack.pushActivity(activity);
    }

    @Override
    public void onActivityStarted(Activity activity) {
        LinkedList<ActivityChangeListener> linkedList = activityLinked.get(activity);
        if (linkedList == null) return;
        for (ActivityChangeListener changeListener : linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.STARTED);
        }
    }

    @Override
    public void onActivityResumed(Activity activity) {
        LinkedList<ActivityChangeListener> linkedList = activityLinked.get(activity);
        if (linkedList == null) return;
        for (ActivityChangeListener changeListener : linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.RESUMED);
        }
    }

    @Override
    public void onActivityPaused(Activity activity) {
        LinkedList<ActivityChangeListener> linkedList = activityLinked.get(activity);
        if (linkedList == null) return;
        for (ActivityChangeListener changeListener : linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.PAUSED);
        }
    }

    @Override
    public void onActivityStopped(Activity activity) {
        LinkedList<ActivityChangeListener> linkedList = activityLinked.get(activity);
        if (linkedList == null) return;
        for (ActivityChangeListener changeListener : linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.STOPPED);
        }
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        ActivityStack.removeActivity(activity);
        LinkedList<ActivityChangeListener> linkedList = activityLinked.get(activity);
        if (linkedList == null) return;
        for (ActivityChangeListener changeListener : linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.DESTROYED);
            changeListener.onActivityDestroy(activity);
        }
        remove(activity);
    }
}

File: activitylifecycle2/src/main/java/com/wghcwc/activitylifecycle2/ActivityLifecycle.kt
package com.wghcwc.activitylifecycle2

import android.app.Activity
import android.app.Application
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import java.util.*

/**
 * @author wghcwc
 * @date 19-12-5
 */
object ActivityLifecycle : ActivityLifecycleCallbacks {
    private val activityLinked = WeakHashMap<Activity, LinkedList<ActivityChangeListener>>()
    private lateinit var app: Application
    private var init = false
    @JvmStatic
    fun init(application: Application) {
        if (init) {
            return
        }
        application.registerActivityLifecycleCallbacks(this)
        app = application
        init = true
    }

    @JvmStatic
    fun getApp(): Application {
        if (!init) {
            throw NullPointerException(" need ActivityLifecycle.init(Application application) in Application")
        }
        return app
    }


    @JvmStatic
    fun add(activity: Activity, listener: ActivityChangeListener?) {
        var linkedList = activityLinked[activity]
        if (linkedList == null) {
            linkedList = LinkedList()
        }
        linkedList.push(listener)
        activityLinked[activity] = linkedList
    }

    @JvmStatic
    fun remove(activity: Activity?, listener: ActivityChangeListener) {
        val linkedList: LinkedList<ActivityChangeListener>? = activityLinked[activity]
        if (linkedList != null) {
            linkedList.remove(listener)
            if (linkedList.size == 0) {
                activityLinked.remove(activity)
            }
        }
    }

    @JvmStatic
    fun remove(activity: Activity?) {
        activityLinked.remove(activity)
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        ActivityStack.pushActivity(activity)
    }

    override fun onActivityStarted(activity: Activity) {
        val linkedList = activityLinked[activity] ?: return
        for (changeListener in linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.STARTED)
        }
    }

    override fun onActivityResumed(activity: Activity) {
        val linkedList = activityLinked[activity] ?: return
        for (changeListener in linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.RESUMED)
        }
    }

    override fun onActivityPaused(activity: Activity) {
        val linkedList = activityLinked[activity] ?: return
        for (changeListener in linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.PAUSED)
        }
    }

    override fun onActivityStopped(activity: Activity) {
        val linkedList = activityLinked[activity] ?: return
        for (changeListener in linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.STOPPED)
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle?) {}
    override fun onActivityDestroyed(activity: Activity) {
        ActivityStack.removeActivity(activity)
        val linkedList = activityLinked[activity] ?: return
        remove(activity)
        for (changeListener in linkedList) {
            changeListener.onActivitySateChange(activity, ActivityState.DESTROYED)
            changeListener.onActivityDestroy(activity)
        }
    }

}